### PR TITLE
pimpl for proposal

### PR DESCRIPTION
**Pull request by victordrobny**
Signed-off-by: Victor Drobny <vdrobny@yandex.ru>

### Description of the Change
Pimpl for proto proposal

### CODE REVIEW

**Comment by l4l** [comment]
Consider fixing CI

File: shared_model/backend/protobuf/impl/proposal.cpp [hunk 1]
```diff
@@ -5,38 +5,58 @@
...
-
-      hash_.invalidate();
-      transactions_.invalidate();
-      blob_.invalidate();
-
-      return *this;
+    struct Proposal::Impl {
+      Impl(TransportType &&ref) : proto_(std::move(ref)) {}
+      Impl(const TransportType &ref) : proto_(ref) {}
+      Impl(Impl &&o) noexcept = delete;
```
Comment (lebdron) [inline shared_model/backend/protobuf/impl/proposal.cpp]: No need to delete move ctor and assignment, as the class is managed by unique_ptr.

File: shared_model/backend/protobuf/proposal.hpp [hunk 2]
```diff
@@ -35,18 +30,16 @@ namespace shared_model {
...
-      template <typename T>
-      using Lazy = detail::LazyInitializer<T>;
+      const TransportType &getTransport() const;
 
-      const Lazy<std::vector<proto::Transaction>> transactions_{[this] {
-        return std::vector<proto::Transaction>(
-            proto_.mutable_transactions()->begin(),
-            proto_.mutable_transactions()->end());
-      }};
+      ~Proposal();
```
Comment (lebdron) [inline shared_model/backend/protobuf/proposal.hpp]: `override` missing.

File: shared_model/backend/protobuf/proposal.hpp [hunk 3]
```diff
@@ -6,25 +6,20 @@
...
+    class Proposal final : public interface::Proposal {
      public:
-      using NonCopyableProto::NonCopyableProto;
+      using TransportType = iroha::protocol::Proposal;
 
       Proposal(Proposal &&o) noexcept;
-      Proposal &operator=(Proposal &&o) noexcept;
+      Proposal &operator=(Proposal &&o) noexcept = default;
+
+      Proposal(TransportType ref);
```
Comment (lebdron) [inline shared_model/backend/protobuf/proposal.hpp]: Please split into "const &" and "&&".

File: shared_model/backend/protobuf/impl/proposal.cpp [hunk 4]
```diff
@@ -5,38 +5,67 @@
...
 
-    Proposal &Proposal::operator=(Proposal &&o) noexcept {
-      proto_ = std::move(o.proto_);
+    Proposal::Proposal(Proposal &&o) = default;
 
-      hash_.invalidate();
-      transactions_.invalidate();
-      blob_.invalidate();
+    Proposal::Proposal(const TransportType &ref) {
+      impl_ = std::make_unique<Proposal::Impl>(std::move(ref));
```
Comment (nickaleks) [inline shared_model/backend/protobuf/impl/proposal.cpp]: std::move is not needed here

File: shared_model/backend/protobuf/impl/proposal.cpp [hunk 5]
```diff
@@ -5,38 +5,67 @@
...
+    }
 
-      return *this;
+    Proposal::Proposal(TransportType &&ref) {
+      impl_ = std::make_unique<Proposal::Impl>(ref);
```
Comment (nickaleks) [inline shared_model/backend/protobuf/impl/proposal.cpp]: std::move is needed here

File: shared_model/backend/protobuf/impl/proposal.cpp [hunk 6]
```diff
@@ -5,38 +5,67 @@
...
 namespace shared_model {
   namespace proto {
     using namespace interface::types;
 
-    Proposal::Proposal(Proposal &&o) noexcept
-        : NonCopyableProto(std::move(o.proto_)) {}
+    struct Proposal::Impl {
+      Impl(TransportType &&ref) : proto_(std::move(ref)) {}
+      Impl(const TransportType &ref) : proto_(ref) {}
+      TransportType proto_;
```
Comment (nickaleks) [inline shared_model/backend/protobuf/impl/proposal.cpp]: add empty lines between members please

File: shared_model/backend/protobuf/impl/proposal.cpp (hunk 5 above)
Comment (l4l) [reply shared_model/backend/protobuf/impl/proposal.cpp]: *std::forward

Comment (lebdron) [reply shared_model/backend/protobuf/impl/proposal.cpp]: TransportType is not a template parameter, so it is std::move

File: shared_model/backend/protobuf/impl/proposal.cpp [hunk 7]
```diff
@@ -5,38 +5,70 @@
...
 namespace shared_model {
   namespace proto {
     using namespace interface::types;
 
-    Proposal::Proposal(Proposal &&o) noexcept
-        : NonCopyableProto(std::move(o.proto_)) {}
+    struct Proposal::Impl {
+      Impl(TransportType &&ref) : proto_(std::move(ref)) {}
+
+      Impl(const TransportType &ref) : proto_(ref) {}
```
Comment (lebdron) [inline shared_model/backend/protobuf/impl/proposal.cpp]: Please use explicit for these two constructors as suggested by codacy.

File: shared_model/backend/protobuf/proposal.hpp [hunk 8]
```diff
@@ -6,25 +6,21 @@
...
-      using NonCopyableProto::NonCopyableProto;
+      using TransportType = iroha::protocol::Proposal;
+
+      Proposal(Proposal &&o);
+      Proposal &operator=(Proposal &&o) noexcept = default;
 
-      Proposal(Proposal &&o) noexcept;
-      Proposal &operator=(Proposal &&o) noexcept;
+      Proposal(const TransportType &ref);
+      Proposal(TransportType &&ref);
```
Comment (lebdron) [inline shared_model/backend/protobuf/proposal.hpp]: Same for codacy and explicit.

File: shared_model/backend/protobuf/proposal.hpp [hunk 9]
```diff
@@ -6,25 +6,21 @@
...
   namespace proto {
-    class Proposal final : public NonCopyableProto<interface::Proposal,
-                                                   iroha::protocol::Proposal,
-                                                   Proposal> {
+    class Proposal final : public interface::Proposal {
      public:
-      using NonCopyableProto::NonCopyableProto;
+      using TransportType = iroha::protocol::Proposal;
+
+      Proposal(Proposal &&o);
```
Comment (lebdron) [inline shared_model/backend/protobuf/proposal.hpp]: Is `noexcept` possible?

**Pull request merged**